docs/06-contributors/999-rfcs/000-stories/story-15.md

```diff
@@ -0,0 +1,186 @@
+but then I noticed that [aws](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-create-api-step-by-step.html) 
+uses the `{id}` syntax. What should be the the right syntax?
+  - We will go with the [openapi spec](https://swagger.io/docs/specification/paths-and-operations/) that uses the `/tasks/{id}` syntax 
+- Bring internal nodejs types - how do we bring something that we can't require, like `RegEx`
+- From anything to Json - can I cast something from bring untyped to Json? 
+- Calling await 
+
+## Code 
+```ts (wing)
+bring cloud;
+bring redis;
+
+// TODO discuss how we bring untyped something like RegEx from JavaScript 
+// PLACEHOLER for bringing something from Javascript stdlib
+bring untyped js ;
+
+// prerequisite: npm install axios
+// PALCEHOLDER for bringing some external module
+bring untyped "axios" as axios; 
+
+enum Status {
+  Uncompleted,
+  Completed
+}
+
+interface ITaskList {
+  inflight get(id: str): Json;
+  inflight add(title: str): str;
+  inflight remove(id: str): void; 
+  inflight find(r: str): Array<str>;
+  inflight set_status(id: str, status: Status): str;
+  inflight set_estimation(id: str, estimation: duration): str;
+}
+
+resource TaskList implementes ITaskList {
+  _redis: redis.Redis;
+  init() {
+    this._redis = new redis.Redis();
+    this.inflights.add("get", ref: "this._redis", op: "GET" );
+    this.inflights.add("_add", ref: "this._redis", op: ["SET", "SADD"]);
+    // notice I am setting explicit permissions on this
+    this.inflights.add("add", ref: "this", op: "_add");
+    // is this the right synatx for multiple ops? 
+    this.inflights.add("set_status", ref: "this", op: ["_add", "get"]); 
+    this.inflights.add("set_estimation", ref: "this", op: ["_add", "get"]); 
+
+    // TODO add more permissions for remove, find
+  }
+
+  inflight get(id: str): Json {
+     return Json.parse(this._redis.GET(id));
+  }
+  
+  inflight _add(id: str, j: Json): str {
+    this._redis.SET(id , Json.format(j));
+    this._redis.SADD("todo", id);
+    return id;
+  } 
+  
+  inflight add(title: str): str {
+    // PLACEHOLDER - how does untyped works with numeric operations
+    let id = "${js.Math.floor(js.Math.random() * 100000000000)}";
+    let j = Json { 
+      title: title, 
+      status: Status.Uncompleted
+    };
+    print("adding task ${id} with data: ${j}"); 
+    return this_add(id, js);
+  }
+
+  inflight remove(id: str) {
+    print("removing task ${id}");
+    this._redis.DEL(id);
+  }
+
+  // PLACEHOLDER - having an untyped type
+  inflight find(term: str): Array<str> { 
+    let r = new js.RegExp(term);
+    let result = MutArray<str>[]; 
+    let ids = this._redis.SMEMBERS("todo");
+    for id in ids {
+      let j = Json.parse(this._redis.GET(id));
+      // Notice that there is autocasting from untyped to bool here 
+      if r.test(j.title) {
+        result.push(id);
+      }
+    }
+    return result.copy();
```

Comment:
I'm still unsure about this API - to me, it feels like converting `MutArray<T>` into `Array<T>` should not require copying data or memory allocation. (It's totally possible I'm not thinking about this right...)

I wonder if it makes sense to say `MutArray<T>` is a subtype of `Array<T>`? From a behavioral perspective, `MutArray<T>` has all of the same behaviors as `Array<T>` (like `get`, `len`, etc.), and includes _extra behaviors_ like `append`, `delete`, etc. If this is the case, then it should be safe to return `result` directly since `MutArray<T>` is a subtype of the annotated return type `Array<str>`.

Thoughts? cc @eladb @yoav-steinberg

Reply:
```
let mutable_array = MutArray<num> [1, 2, 3];

// If MutArray is allowed to be assigned like this without cloning
let immutable_array: Array<num> = mutable_array;

// immutable_array has changed, thus defeating the point of it being immutable
mutable_array.push(4)
```

Reply:
🤦‍♂️  right, I see... (great example btw)

It's sort of like a 3 line math proof, and the question whether the step from line 1 to 2 is wrong, or the step from line 2 to 3 is wrong.

In Rust, step 2 to 3 is invalid because `mutable_array` gets "moved" to immutable_array, so it can't be used by the old variable anymore. But since we don't do that kind of static analysis in our language (doesn't really make sense I think?) then step 2 to 3 must be OK. Thus the subtyping between steps 1 and 2 is wrong for our language.

So I guess as a rule of thumb, to the extent that our language has subtyping, in order to figure out whether Y is a subtype of X, we have to consider whether the relationship breaks any "mutation" invariants, and not just whether Y inherits the same set of behaviors. (at least that's my initial takeaway)

Reply:
In that case `result.copy()` makes sense, so I'm resolving this thread 👍